### PR TITLE
PLAT-101893: Add ImageItem

### PR DESCRIPTION
**Pull request by MikyungKim**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [ ] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Need to support vertical and horizontal image item

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
Added ImageItem with `orientation` prop

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
PLAT-101893

### CODE REVIEW

File: packages/ui/ImageItem/ImageItem.js [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+				subComponents ? subComponents : <React.Fragment>
+					{caption ? (<Cell className={css.caption} component={Caption} shrink>{caption}</Cell>) : null}
+					{subCaption ? (<Cell className={css.subCaption} component={Caption} shrink>{subCaption}</Cell>) : null}
+				</React.Fragment>
+			);
+		}
+	},
+
+	render: ({className, css, imageComponent: ImageComponent, orientation, placeholder, source, selectionOverlay, subComponents, ...rest}) => {
+		const classes = classNames(className, orientation === 'vertical' ? css.vertical : css.horizontal);
```
Comment (ybsung) [inline packages/ui/ImageItem/ImageItem.js]: Why don't you manipulate the class names in the `computed` block in a kind?
```
	computed: {
		className: ({orientation, styler}) => styler.append({vertical: , horizontal: })
	},
```

File: packages/ui/ImageItem/ImageItem.js [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+				</Cell>
+				{subComponents}
+			</Column> : <Row {...rest} className={classes} inline>
+				<Cell className={css.image} component={ImageComponent} placeholder={placeholder} shrink src={source}>
+					{selectionOverlay}
+				</Cell>
+				<Cell align="center">
+					{subComponents}
+				</Cell>
+			</Row>
```
Comment (ybsung) [inline packages/ui/ImageItem/ImageItem.js]: How about simplifying render function by wrapping subComponents with Cell in `computed` block?

```
	computed: {
		...
		subComponents: ({caption, captionComponent: Caption, css, orientation, subCaption, subComponents}) => {
			const Component = (
				subComponents ? subComponents : <React.Fragment>
					{caption ? (<Cell className={css.caption} component={Caption} shrink>{caption}</Cell>) : null}
					{subCaption ? (<Cell className={css.subCaption} component={Caption} shrink>{subCaption}</Cell>) : null}
				</React.Fragment>
			);

			return orientation === 'horizontal' ? 
				<Cell align="center" component={Component} /> :
				component;
		}
	},

	render: (...) => {
		...
		const isHorizontal= orientation === 'horizontal';
		const Component = isHorizontal ? Row : Column;

		return (
			<Component {...rest} className={classes} inline>
				<Cell className={css.image} component={ImageComponent} placeholder={placeholder} shrin={isHorizontal} src={source}>
					{selectionOverlay}
				</Cell>
				{subComponents}
			</Component>
		);
	}
```

File: packages/ui/ImageItem/ImageItem.js (hunk 1 above)
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: oh yeah, good point.

File: packages/ui/ImageItem/ImageItem.js (hunk 2 above)
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: I guess so... It looks simpler but doesn't seem good to read, though.

File: packages/ui/ImageItem/ImageItem.js [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	styles: {
+		css: componentCss,
+		className: 'imageItem',
+		publicClassNames: true
+	},
+
+	computed: {
+		className: ({css, orientation, selected, styler}) => styler.append(
+			{selected, horizontal: orientation === 'horizontal' ? css.horizontal : null, vertical: orientation === 'vertical' ? css.vertical : null}
+		),
```
[the remark this answers is not in the document]
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: oh good point!

File: packages/ui/ImageItem/ImageItem.js [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		 * 	render: () => <div>custom overlay</div>
+		 * });
+		 *
+		 * <ImageItem selectionOverlay={SelectionOverlay} />
+		 * ```
+		 *
+		 * @type {Function}
+		 * @public
+		 */
+		selectionOverlay: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: okay

File: packages/ui/ImageItem/ImageItem.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+		/**
+		 * Shows a selection overlay with a centered icon. When `selected` is true, a check mark is shown.
+		 *
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		selectionOverlayShowing: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: Sure. That sounds more general.

File: packages/ui/ImageItem/ImageItem.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+		/**
+		 * The absolute URL path to the image.
+		 *
+		 * @type {String}
+		 * @public
+		 */
+		source: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (MikyungKim) [reply packages/ui/ImageItem/ImageItem.js]: 👍

**Comment by MikyungKim** [comment]
@ryanjduffy Thanks for the review. Actually, I was trying not to break the compatibility with current ui/GridListImageItem API since moonstone/GridListImageItem is using it. If we don't care about breaking API compatibility with moonstone after 4.0.0, I think we could make ui/ImageItem more general.

I also think the
- an image and a placeholder for that image,
- a caption,
- selected
would be the minimum.

I'll try what I can do and let you know when I need help. Thanks!

**Comment by ryanjduffy** [comment]
We do need to be concerned with moonstone but it would continue to use `GridListImageItem` which is deprecated but still supported. If we can't adapt `ui/GridListImageItem` to use the new `ui/ImageItem`, we'd have to live with the duplication until 4.0

**Pull request merged**